src/compile.rs

```diff
@@ -682,6 +706,131 @@ fn check_required_config<'a>(
     ret.into_result(())
 }
 
+// Get the types, functions, and policy rules for the system.
+pub fn get_reduced_infos(
+    policies: &[PolicyFile],
+    classlist: &ClassList,
+    system: &ValidatedSystem,
+    type_map: &TypeMap,
+    module_map: &ModuleMap,
+) -> Result<Vec<sexp::Sexp>, CascadeErrors> {
+    let ret = CascadeErrors::new();
+    let mut new_type_map = get_built_in_types_map()?;
+
+    // Get the reduced type infos
+    for module in &system.modules {
+        get_reduced_types(module, &mut new_type_map, type_map, module_map)?;
+    }
+
+    // Generate type aliases for the new reduced type map
+    let new_t_aliases = collect_aliases(new_type_map.iter());
+    new_type_map.set_aliases(new_t_aliases);
+
+    // Get the function infos
+    let mut new_func_map = get_funcs(policies, &new_type_map)?;
+
+    // Validate functions
+    let new_func_map_copy = new_func_map.clone(); // In order to read function info while mutating
+    validate_functions(
+        &mut new_func_map,
+        &new_type_map,
+        classlist,
+        &new_func_map_copy,
+    )?;
+
+    // Get the policy rules
+    let new_policy_rules = get_policy_rules(policies, &new_type_map, classlist, &new_func_map)?;
+
+    let new_cil_tree = generate_sexp(
+        &new_type_map,
+        classlist,
+        new_policy_rules,
+        &new_func_map,
+        &Some(&system.configurations),
+    )?;
+
+    ret.into_result(new_cil_tree)
+}
+
+// This is a recusive function that gets only the relevant types from the type map.
+// The reduced types are the types in the module and the types in any of that modules' child modules.
+// Parents of those types are also automatically included.
+// Types are cloned for ownership.
+pub fn get_reduced_types(
+    module: &ValidatedModule,
+    reduced_type_map: &mut TypeMap,
+    type_map: &TypeMap,
+    module_map: &ModuleMap,
+) -> Result<(), CascadeErrors> {
+    for t in &module.types {
+        if let Some(type_info) = type_map.get(t.name.as_ref()) {
+            if !reduced_type_map.iter().any(|(k, _v)| k == t.name.as_ref()) {
+                reduced_type_map.insert(t.name.to_string(), type_info.clone())?;
+            }
+        }
+        for parent in &t.inherits {
+            if let Some(parent_type_info) = type_map.get(parent.as_ref()) {
+                if !reduced_type_map.iter().any(|(k, _v)| k == parent.as_ref()) {
+                    reduced_type_map.insert(parent.to_string(), parent_type_info.clone())?;
+                }
+            }
+        }
+    }
+    for vm in &module.validated_modules {
+        if let Some(child_module) = module_map.get(vm.as_ref()) {
+            get_reduced_types(child_module, reduced_type_map, type_map, module_map)?;
+        }
+    }
+    Ok(())
+}
+
+pub fn get_funcs<'a>(
```

Comment:
Putting this here for lack of a better place...

So, we reduce the function map.  Which means that theoretically, some function (call it foo.my_func) may exist in the policy, but not in the particular system being compiled.  So when I inside bar (which is part of my system) call foo.my_func, then it's no longer in the map.

I'm assuming that in that case, we'd get a "this function doesn't exist message" (hopefully it's that rather than generating invalid policy that calls the function but doesn't declare it?)

If that's all accurate, that's fine for now, but we should capture the current behavior somewhere and make a note while we're in the code and understanding everything how to update it later.  I *think* that in that case, we'll generally be wanting to silently drop those calls rather than returning an error.  That should probably get captured as a TODO for later, since I still need to write up some more formal design for it and discuss with people.

Am I missing anything here?  Can you figure out where to document the status of this particular aspect of compilation and document it please?